ggml/src/ggml-backend.cpp

```diff
@@ -407,170 +426,239 @@ void ggml_backend_event_wait(ggml_backend_t backend, ggml_backend_event_t event)
+}
 
-#ifdef GGML_USE_VULKAN
-    extern GGML_CALL int ggml_backend_vk_reg_devices(void);
-    ggml_backend_vk_reg_devices();
-#endif
+bool ggml_backend_dev_offload_op(ggml_backend_dev_t device, const struct ggml_tensor * op) {
+    return device->iface.offload_op(device, op);
+}
 
-#ifdef GGML_USE_KOMPUTE
-    extern GGML_CALL void ggml_backend_kompute_reg_devices(void);
-    ggml_backend_kompute_reg_devices();
-#endif
+ggml_backend_event_t ggml_backend_dev_event_new(ggml_backend_dev_t device) {
+    if (!device->iface.event_new) {
+        return NULL;
+    }
+    return device->iface.event_new(device);
+}
 
-#ifdef GGML_USE_CANN
-    extern GGML_CALL int ggml_backend_cann_reg_devices(void);
-    ggml_backend_cann_reg_devices();
-#endif
+// Backend (reg)
+
+const char * ggml_backend_reg_name(ggml_backend_reg_t reg) {
+    return reg->iface.get_name(reg);
 }
 
-GGML_CALL void ggml_backend_register(const char * name, ggml_backend_init_fn init_fn, ggml_backend_buffer_type_t default_buffer_type, void * user_data) {
-    GGML_ASSERT(ggml_backend_registry_count < GGML_REG_MAX_BACKENDS);
+size_t ggml_backend_reg_dev_count(ggml_backend_reg_t reg) {
+    return reg->iface.device_count(reg);
+}
 
-    size_t id = ggml_backend_registry_count;
+ggml_backend_dev_t ggml_backend_reg_dev_get(ggml_backend_reg_t reg, size_t index) {
+    return reg->iface.device_get(reg, index);
+}
 
-    ggml_backend_registry[id] = (struct ggml_backend_reg) {
-        /* .name                = */ {0},
-        /* .fn                  = */ init_fn,
-        /* .default_buffer_type = */ default_buffer_type,
-        /* .user_data           = */ user_data,
-    };
+void * ggml_backend_reg_get_proc_address(ggml_backend_reg_t reg, const char * name) {
+    if (!reg->iface.get_proc_address) {
+        return NULL;
+    }
+    return reg->iface.get_proc_address(reg, name);
+}
 
-    snprintf(ggml_backend_registry[id].name, sizeof(ggml_backend_registry[id].name), "%s", name);
+void ggml_backend_reg_set_log_callback(ggml_backend_reg_t reg, ggml_log_callback log_callback, void * user_data) {
+    if (reg->iface.set_log_callback) {
+        reg->iface.set_log_callback(reg, log_callback, user_data);
+    }
+}
 
-#ifndef NDEBUG
-    fprintf(stderr, "%s: registered backend %s\n", __func__, name);
+// Backend registry
+
+#ifdef GGML_USE_CUDA
+#include "ggml-cuda.h"
 #endif
 
-    ggml_backend_registry_count++;
-}
+struct ggml_backend_registry {
+    std::vector<ggml_backend_reg_t> backends;
+    std::vector<ggml_backend_dev_t> devices;
 
-size_t ggml_backend_reg_get_count(void) {
-    ggml_backend_registry_init();
+    ggml_backend_registry() {
+#ifdef GGML_USE_CUDA
+        register_backend(ggml_backend_cuda_reg());
+#endif
 
-    return ggml_backend_registry_count;
-}
+        register_backend(ggml_backend_cpu_reg());
 
-size_t ggml_backend_reg_find_by_name(const char * name) {
-    ggml_backend_registry_init();
+        // TODO: sycl, metal, vulkan, kompute, cann
+    }
```

Comment:
Is there a meaning behind the order of backends, e.g. the priority with which they are used?

Reply:
Functions like `ggml_backend_dev_by_type` choose the first device of the given type, so the order can make a different.